src/text-editor-presenter.coffee

```diff
@@ -298,50 +302,46 @@ class TextEditorPresenter
     @state.content.backgroundColor = if @model.isMini() then null else @backgroundColor
     @state.content.placeholderText = if @model.isEmpty() then @model.getPlaceholderText() else null
 
-  updateLinesState: ->
+  updateTilesState: ->
     return unless @startRow? and @endRow? and @lineHeight?
 
-    visibleLineIds = {}
-    row = @startRow
-    while row < @endRow
-      line = @model.tokenizedLineForScreenRow(row)
-      unless line?
-        throw new Error("No line exists for row #{row}. Last screen row: #{@model.getLastScreenRow()}")
-
-      visibleLineIds[line.id] = true
-      if @state.content.lines.hasOwnProperty(line.id)
-        @updateLineState(row, line)
+    linesPerTile = Math.floor(@height / @lineHeight / @tileCount)
+    linesPerTile = Math.max(1, linesPerTile)
+
+    startIndex = Math.max(
+      0, Math.floor(@startRow / linesPerTile) - @tileOverdrawMargin
+    )
+    endIndex = Math.min(
+      Math.ceil(@model.getScreenLineCount() / linesPerTile),
+      Math.ceil(@endRow / linesPerTile) + @tileOverdrawMargin
+    )
+
+    visibleTiles = {}
+    for index in [startIndex...endIndex]
+      presenter = @linesPresentersByTileIndex[index] ?= new LinesPresenter(@)
```

Comment:
I have inlined it for the time being :fire: 

The main idea with having a separate object, though, was to push the design of presenters/components in another direction. This object, as it is right now, tends to couple different behaviors in the same place and I can feel a lot of cognitive overhead every time I add new features or fix :bug:s (e.g. when I change how lines construction behave I have to think about the details of how cursors work, and vice versa).

:bulb: By separating each responsibility, I think we could be able to understand each presenter at a glance and, as a result, feel more confident to change it.

Probably I have been a bit too abstract here, but I'd love to hear your opinion about moving towards more cohesive presenters in the future :thought_balloon: